MindSphere Improvements (#1584)
mendix/docs: content/developerportal/deploy/deploying-to-mindsphere.md

title: "Siemens MindSphere"
category: "Deployment"
menu_order: 45

> description: "Describes how to deploy a Mendix app to the MindSphere Gateway"

Reviewer: Hmm, actually you 
- register apps at the MindSphere Gateway
- integrate apps into the MindSphere Launchpad
- deploy apps to MindSphere

Reviewer: ```suggestion
description: "Describes how to register a Mendix app at the MindSphere Gateway and integrate it into the MindSphere Launchpad"
```

title: "Siemens MindSphere"
category: "Deployment"
menu_order: 45
description: "Describes how to deploy a Mendix app to MindGate, the MindSphere launchpad"

> tags: ["MindSphere", "deploy", "cloud foundry", "launchpad", "scopes", "roles", "sso", "XSRF", "limitations", "MindGate"]

Reviewer 2: Can anyone make a comment?

Author: Yes

{{% /alert %}}

{{% alert type="warning" %}}
There are some limitations to what you can do in your Mendix app if it is deployed to MindSphere. See the [Limitations](/refguide/mindsphere/mindsphere-development-considerations#limitations) section of *MindSphere Development Considerations* for more information.
{{% /alert %}}

{{% alert type="info" %}}
You can easily copy code examples shown within grey blocks into the clipboard. Hover the cursor over the code block and click the copy button which appears.

![](attachments/deploying-to-mindsphere/copy-from-documentation.png)

{{% /alert %}}

To help you with your first MindSphere apps, there is also an example app which contains modules which call the MindSphere APIs. See [How to Use the Siemens MindSphere Pump Asset Example App](/howto/mindsphere/mindsphere-example-app) for more information.

## 2 Prerequisites{#prerequisites}

To deploy your app to MindSphere you need the following prerequisites.

* A MindSphere user account on a developer tenant

> * The Cloud Foundry Command Line Interface (CF CLI) – this can be downloaded from [https://github.com/cloudfoundry/cli](https://github.com/cloudfoundry/cli)

Reviewer: ```suggestion
* [Mendix Desktop Modeler](https://appstore.home.mendix.com/index3.html)
* The Cloud Foundry Command Line Interface (CF CLI) – this can be downloaded from [https://github.com/cloudfoundry/cli](https://github.com/cloudfoundry/cli)
```

* A Cloud Foundry role which allows you to push applications, such as `SpaceDeveloper` (help in setting up Cloud Foundry users can be found in the MindSphere [Cloud Foundry How Tos](https://developer.mindsphere.io/paas/paas-cloudfoundry-howtos.html))
* A MindSphere developer role: either `mdsp:core:Developer` or `mdsp:core:DeveloperAdmin`

## 3 Including Required MindSphere Modules

You must customize your app to allow it to be deployed to MindSphere, registered via the MindSphere Developer Cockpit, and shown in the launchpad. This is done through MindSphere customization modules. There are two ways to include the customization you need in your app.

> ### 3.1 Using the MindSphere Starter App

Reviewer: ```suggestion
### 3.1 Option A: Using the MindSphere Starter App
```

The **MindSphere Starter Application** in the Mendix App Store contains all the modules and styling which you need to create an app you want to deploy to MindSphere.

{{% alert type="info" %}}
This is the recommended approach if you are building a new application, as it will provide all the necessary building blocks to get started.
{{% /alert %}}

Open the Desktop Modeler (version 7.22.2 or above) and follow these steps:

1. Click the icon in the top-right of the menu bar to open the Mendix App Store.

	![](attachments/deploying-to-mindsphere/app-store-icon.png)

2. Enter *MindSphere* in the search box, and press <kbd>Enter</kbd>.

3. Select **MindSphere Starter Application** in the search results.

	![](attachments/deploying-to-mindsphere/app-store-search.png)
  
4. Click **Download** to create a new app project using this app.

	![](attachments/deploying-to-mindsphere/app-store-download.png)
  
5. To start the new app project, confirm where to store the app, the app name, and the project directory, then click **OK**.

	![](attachments/deploying-to-mindsphere/app-store-download-project.png)

> ### 3.2 Customizing an Existing App{#existingapp}

Reviewer: ```suggestion
### 3.2 Option B: Customizing an Existing App{#existingapp}
```

  This integrates the mandatory MindSphere OS Bar with your app. For more information, see the [MindSphere OS Bar](/refguide/mindsphere/mindsphere-module-details#msosbar) section of *MindSphere Module Details*.

* MindSphere Theme Pack (MindSphere_UI_Resources) from the Mendix App Store here: [Siemens MindSphere Theme Pack](https://appstore.home.mendix.com/link/app/108803/)

  This applies MindSphere styling to your app and includes some additional custom files which are required for the correct operation of your app. For more information, see the [MindSphere Theme Pack](/refguide/mindsphere/mindsphere-module-details#msthemepack) section of *MindSphere Module Details*.

## 4 Configuring the Modules

Now that you have your new app, or have imported the MindSphere modules into an existing app, you need to configure the modules to allow your app to work with MindSphere.

### 4.1 Configuring Single Sign-On (MindSphereSingleSignOn)

The following items in the MindSphereSingleSignOn module need to be configured.

![Folder structure of the MindSphereSingleSignOn module](attachments/deploying-to-mindsphere/image2.png)

> #### 4.1.1 CockpitApplicationName

Reviewer: Could you restructure this to have a numbered list of actions instead of descriptions? That would make it easier for users to follow and to tell whether they have made all necessary configurations.

Author: The actions were already numbered as sections of the document.

I have started each section with an imperative to make it clear what has to be done. Hope that works OK when it is published.

Reviewer: Oh, I didn't interpret the section numbers as action items. Maybe because the headline is not phrased as imperative. Anyway, let's publish first an keep this in the backlog

3.  Create a PostgreSQL instance using the command:

    ```bash
    cf create-service postgresql10 {plan} {service_instance} [-c {parameters_as_JSON}] [-t {tags}]
    ```

    For example: `cf create-service postgresql10 postgresql-xs myapp-db`  

    For more information see [Using the a9s PostgreSQL](https://developer.mindsphere.io/paas/a9s-postgresql/using.html) on the MindSphere developers site.

4.  Depending on your infrastructure and service broker usage, it may take several minutes to create the service instance. Check if your PostgreSQL service has been created successfully using the following command:  
    `cf services`  
    Your service should be listed, and the last operation should be ‘create succeeded’.

5.  Ensure you are in the same folder as the package you wish to deploy.

6. Create a `manifest.yml` file with at least the following content:

    ```yml
    applications:
    - name: {app_name}
      disk_quota: 512M

>       memory: 512M

Reviewer: please replace the values for disk_quota and memory by placeholders and add in info box about the minimum requirements

Author: Done

10.  Specify `/**` as the endpoint to allow you to access all endpoints relevant to your application, and click **Save**.

11. Fill in the **Cloud Foundry Direct URL**. This can be found using the cloud foundry command `cf app {app_name}`.

12.  Set the **Configurations > content-security-policy** *Value* to the following:

      ```http
      default-src 'self' 'unsafe-inline' 'unsafe-eval' static.eu1.mindsphere.io sprintr.home.mendix.com; font-src 'self' static.eu1.mindsphere.io fonts.gstatic.com; style-src * 'unsafe-inline'; script-src 'self' 'unsafe-inline' 'unsafe-eval' static.eu1.mindsphere.io sprintr.home.mendix.com; img-src * data:;
      ```

      {{% alert type="info" %}}These content security policy (CSP) settings are needed to ensure that the MindSphere OS Bar and the [Mendix Feedback Widget](https://appstore.home.mendix.com/link/app/199/) are loaded correctly. You may need to set additional CSP settings if you make additional calls to other domains (for example, if you use Google maps from maps.googleapi.com).{{% /alert %}}

      ![](attachments/deploying-to-mindsphere/image14.png)

13.  Click **Save** to save these details.

> 14.  Click **Register** to register your app with the MindSphere launchpad.

Reviewer: Maybe start a new section **Registering the App** here to be more consistent with the MindSphere documentation? We should at least have similar headlines and structure as those can be compared easily without going through all the details.

The user will have to log out and log in again for this assignment to take effect.
{{% /alert %}}

{{% alert type="success" %}}
Your app is now set up and users can run it from within the MindSphere Developer Cockpit.
{{% /alert %}}

> ## 6 Development Considerations

Reviewer: ```suggestion
## 6 Related Links
```

Author: I'd rather leave this as it is to improve searching and navigation. I'll have another look when I've published the document.

> 

Author: If needed we can discuss this on Thursday.
The process I'm referring to here is the one documented in https://code.siemens.com/mindsphere-mainline/documentation/blob/master/docs/howto/text-modules/login-cf.txt
This isn't the same as going to the MindSphere launchpad.
I'll add the steps from your document/the howto-cf-running-app document since there are only three steps.

> 

Reviewer: The content of https://code.siemens.com/mindsphere-mainline/documentation/blob/master/docs/howto/text-modules/login-cf.txt has nothing to do with the application name, so please add it to the discussion items for tomorrow.

Affected section:
```markdown
**CockpitApplicationName**
This is the name of your app as registered in the MindSphere developer portal. See [Running a Cloud Foundry-Hosted Application](https://developer.mindsphere.io/howto/howto-cf-running-app.html#configure-the-application-via-the-developer-cockpit) for more information.
```

> 

Author: @rikekroegersiemens - I've added this to the description so we can use that as an agenda. Feel free to add other items.

![Domain model showing MindSphereToken entity](attachments/deploying-to-mindsphere/image5.png)

> If the access token can be retrieved from the environment, this is used. If a valid token cannot be retrieved, *and the app is running locally*, then the user is asked to sign on by providing their credentials manually. This enables the app to be tested locally, without having to be deployed to the MindSphere environment after every change. You should check whether the access token has been successfully retrieved using the query `${MindSphereTokenName} != empty`. For example, `$MindSphereToken != empty` in the scenario shown in the image below.

Reviewer: In that case, maybe remove the first sentence completely?

#### 6.3.5  No Authorization Information Found Page

The *No Authorization Information Found* page (NoJWT.html) will be shown if your app is called without a valid token. This happens when the app URL is called directly and not via MindGate (launchpad). For example, if you have a self-hosted, or Mendix Cloud, deployment. The SSO module expects to find this MindSphere-compliant file as error_page/NoJWT.html within your ‘Theme’ folder.

#### 6.3.6  CockpitApplicationName Not Found in the Provided Authorization Information Page

The *CockpitApplicationName does not match MindSphere token* page will be shown if your app is called with a token which does not include the value (as the *audience* claim of the JWT) you have specified within the SSO constant ‘CockpitApplicationName’. The SSO module expects to find this MindSphere-compliant file as error_page/CockpitApplicationName.html within your ‘Theme’ folder.

## 7 Development Considerations

See [MindSphere Development Considerations](/refguide/mindsphere/mindsphere-development-considerations) for additional help on such things as:

* local testing
* multi-tenancy
* limitations

## 8 Appendices

### 8.1 index.html{#indexhtml}

> Various changes need to be made to the standard Mendix index.html file to ensure compatibility with MindSphere.

Reviewer: I can't see the changes yet
